test/test_model.py

```diff
@@ -1721,8 +1722,40 @@ def test_action_get(self):
     def test_action_set(self):
         fake_script(self, 'action-get', 'exit 1')
         fake_script(self, 'action-set', 'exit 0')
-        self.backend.action_set(OrderedDict([('x', 'dead beef'), ('y', 1)]))
-        self.assertEqual(fake_script_calls(self), [['action-set', 'x=dead beef', 'y=1']])
+        self.backend.action_set({'x': 'dead beef', 'y': 1})
+        expected_args = ['x=dead beef', 'y=1']
+        self.assertTrue(all(item in fake_script_calls(self)[0] for item in expected_args))
+
+    def test_action_set_key_validation(self):
+        with self.assertRaises(ValueError):
+            self.backend.action_set({'X': 'dead beef', 'y': 1})
+        with self.assertRaises(ValueError):
+            self.backend.action_set({'some&key': 'dead beef', 'y': 1})
+        with self.assertRaises(ValueError):
+            self.backend.action_set({'someKey': 'dead beef', 'y': 1})
+        with self.assertRaises(ValueError):
+            self.backend.action_set({'some_key': 'dead beef', 'y': 1})
+
+    def test_action_set_nested(self):
+        fake_script(self, 'action-get', 'exit 1')
+        fake_script(self, 'action-set', 'exit 0')
+        self.backend.action_set({'a': {'b': 1, 'c': 2}, 'd': 3})
+        expected_args = ['a.b=1', 'a.c=2', 'd=3']
+        self.assertTrue(all(item in fake_script_calls(self)[0] for item in expected_args))
+
+    def test_action_set_more_nested(self):
+        fake_script(self, 'action-get', 'exit 1')
+        fake_script(self, 'action-set', 'exit 0')
+        self.backend.action_set({'a': {'b': 1, 'c': 2, 'd': {'e': 3}}, 'f': 4})
+        expected_args = ['a.b=1', 'a.c=2', 'a.d.e=3', 'f=4']
+        self.assertTrue(all(item in fake_script_calls(self)[0] for item in expected_args))
+
+    def test_action_set_dotted_dict(self):
+        fake_script(self, 'action-get', 'exit 1')
+        fake_script(self, 'action-set', 'exit 0')
+        self.backend.action_set({'a.b': 1, 'a': {'c': 2}, 'd': 3})
```

Comment:
I'm not sure we should accept `.`s in keys, as it could lead to problematic situations. For example, what if you pass in `{'a.b': 1, 'a': {'b': 2}, 'd': 3}`? Alternatively we could just raise an exception if such a conflict is found.

Reply:
Yeh interesting. I initially didn't allow the `.`, but that is likely to break existing charms that are using `set_result`, so I think it's important to allow it.

In the case you mention, you'd end up with `{'a.b': 2, 'd': 3}`.

Reply:
Yeah, I still think that might be quite unexpected, and we should likely throw an exception if there's a collision like this rather than silently overwriting/discarding one value.

Reply:
Yeh, though I guess it's less surprising given that it's the default behaviour for dicts with duplicate keys:

```
❯ python3
Python 3.9.5 (default, May 11 2021, 08:20:37) 
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> {'a.b': 1, 'a.b': 2, 'd': 3}
{'a.b': 2, 'd': 3}
>>> 
```

This feels (to me at least) like it would be more an error in charm code that should be caught with unit tests?

Reply:
I did get an implementation that checks for this, but I prefer John's approach below. Have implemented in the latest commit.